internal/graph/memory.go

```diff
@@ -101,34 +87,71 @@ func (m *Memory) IndexAll(ctx context.Context, fetcher content.Fetcher, node oci
 // it does not necessarily correspond to any consistent snapshot of the stored
 // contents.
 func (m *Memory) Predecessors(_ context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
 	key := descriptor.FromOCI(node)
-	value, exists := m.predecessors.Load(key)
+	set, exists := m.predecessors[key]
 	if !exists {
 		return nil, nil
 	}
-	predecessors := value.(*sync.Map)
-
 	var res []ocispec.Descriptor
-	predecessors.Range(func(key, value interface{}) bool {
-		res = append(res, value.(ocispec.Descriptor))
-		return true
-	})
+	for k := range set {
+		res = append(res, m.nodes[k])
+	}
 	return res, nil
 }
 
+// Remove removes the node from its predecessors and successors.
+func (m *Memory) Remove(ctx context.Context, node ocispec.Descriptor) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	nodeKey := descriptor.FromOCI(node)
+	// remove the node from its successors' predecessor list
+	for successorKey := range m.successors[nodeKey] {
+		predecessorEntry := m.predecessors[successorKey]
+		predecessorEntry.Delete(nodeKey)
+		// remove the successorKey entry from m. predecessors, if it's an empty set.
+		// To it another way, if all its predecessors are already deleted, then we
+		// delete the entry in m.predecessors. In any of its predecessors still exists,
+		// the entry is NOT deleted. The presence of the predecessors entry depends on
+		// the status of all predecessors of the node.
```

Comment:
nit: I'm a bit of lost when reading this😂

Reply:
Rephrased it as the following: `if none of the predecessors of the node still exists, we remove the predecessors entry. Otherwise, we do not remove the entry.`